src/mlpack/methods/reinforcement_learning/q_networks/dueling_dqn.hpp

```diff
@@ -0,0 +1,246 @@
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_RL_DUELING_DQN_HPP
+#define MLPACK_METHODS_RL_DUELING_DQN_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <mlpack/methods/ann/ffn.hpp>
+#include <mlpack/methods/ann/init_rules/gaussian_init.hpp>
+#include <mlpack/methods/ann/layer/layer.hpp>
+#include <mlpack/methods/ann/loss_functions/mean_squared_error.hpp>
+#include <mlpack/methods/ann/loss_functions/empty_loss.hpp>
+
+namespace mlpack {
+namespace rl {
+
+using namespace mlpack::ann;
+
+/**
+ * @tparam CompleteNetworkType The type of network used for full dueling dqn.
+ * @tparam FeatureNetworkType The type of network used for feature network.
+ * @tparam AdvantageNetworkType The type of network used for advantage network.
+ * @tparam ValueNetworkType The type of network used for value network.
+ */
+template <
+  typename CompleteNetworkType = FFN<EmptyLoss<>, GaussianInitialization>,
+  typename FeatureNetworkType = Sequential<>,
+  typename AdvantageNetworkType = Sequential<>,
+  typename ValueNetworkType = Sequential<>
+>
+class DuelingDQN
+{
+ public:
+  /**
+   * Default constructor.
+   */
+  DuelingDQN()
+  { /* Nothing to do here. */ }
+
+  //! Copy constructor.
+  DuelingDQN(const DuelingDQN& model):
+      featureNetwork(model.featureNetwork),
+      advantageNetwork(model.advantageNetwork),
+      valueNetwork(model.valueNetwork),
+      actionValues(model.actionValues)
+  {
+    std::cout << "copyConstructor" << '\n';
+    
+    concat = new Concat<>(true);
+    concat.Add(&valueNetwork);
+    concat.Add(&advantageNetwork);
+
+    completeNetwork.Add(new IdentityLayer<>());
+    completeNetwork.Add(&featureNetwork);
+    completeNetwork.Add(&concat);
+    completeNetwork.Parameters() = model.completeNetwork.Parameters();
+  };
+
+  //! Move constructor.
+  DuelingDQN(DuelingDQN&& model):
+      featureNetwork(std::move(model.featureNetwork)),
+      advantageNetwork(std::move(model.advantageNetwork)),
+      valueNetwork(std::move(model.valueNetwork)),
+      actionValues(std::move(model.actionValues))
+  {
+    std::cout << "move Constructor" << '\n';
+
+    concat = new Concat<>(true);
+    concat.Add(&valueNetwork);
+    concat.Add(&advantageNetwork);
+
+    completeNetwork.Add(new IdentityLayer<>());
+    completeNetwork.Add(&featureNetwork);
+    completeNetwork.Add(&concat);
+    completeNetwork.Parameters() = model.completeNetwork.Parameters();
+  }
+
+  DuelingDQN& operator = (DuelingDQN model)
+  {
+    std::swap(this->completeNetwork, model.completeNetwork);
+    std::swap(featureNetwork, model.featureNetwork);
+    std::swap(valueNetwork, model.valueNetwork);
+    std::swap(advantageNetwork, model.advantageNetwork);
+    return *this;
+  };
```

Comment:
Hey!
I had been trying to get a clone for the complete network, required in [here](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/reinforcement_learning/q_learning_impl.hpp#L257) and [here](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/reinforcement_learning/q_learning_impl.hpp#L65).
But I keep getting an error that `boost::detail::variant::forced_return() [with T = void]: Assertion `false' failed.`. My first guess was that since the network deals with pointers, so I need to make copy and move constructors, and then properly link all the networks. But I dont think that works either. 

Here's the code to reproduce the error.
```c++
using namespace mlpack;
using namespace mlpack::ann;
using namespace ens;
using namespace mlpack::rl;

int main()
{
  mlpack::math::RandomSeed(std::time(NULL));

  // Set up the policy and replay method.
  GreedyPolicy<CartPole> policy(1.0, 1000, 0.1, 0.99);
  RandomReplay<CartPole> replayMethod(10, 100);

  TrainingConfig config;
  config.StepSize() = 0.01;
  config.Discount() = 0.9;
  config.TargetNetworkSyncInterval() = 100;
  config.ExplorationSteps() = 100;
  config.DoubleQLearning() = false;
  config.StepLimit() = 200;

  DuelingDQN<> network(4, 128, 64, 2);

// TO CHECK FORWARD AND BACKWARD PASSES
  // arma::mat input = {1,2,3,4};
  // arma::mat target = {1,2};
  // input = input.t();
  // target = target.t();
  // arma::mat output, gradient;
  // network.Forward(input, output);
  // std::cout << output;
  // network.Backward(input, target, gradient);
  // std::cout << gradient;

  // Set up DQN agent.
  QLearning<CartPole, decltype(network), AdamUpdate, decltype(policy)>
      agent(std::move(config), std::move(network), std::move(policy),
            std::move(replayMethod));

  arma::running_stat<double> averageReturn;
  size_t episodes = 0;
  bool converged = true;
  while (true)
  {
    double episodeReturn = agent.Episode();
    averageReturn(episodeReturn);
    episodes += 1;

    if (episodes > 1000)
    {
      std::cout << "Cart Pole with DQN failed." << std::endl;
      converged = false;
      break;
    }
    std::cout << "Average return: " << averageReturn.mean()
              << " Episode return: " << episodeReturn << std::endl;
    if (averageReturn.mean() > 100)
    {
      agent.Deterministic() = true;
      arma::running_stat<double> testReturn;
      for (size_t i = 0; i < 10; ++i)
        testReturn(agent.Episode());

      std::cout << "Average return in deterministic test: "
                << testReturn.mean() << std::endl;
      break;
    }
  }
}
```
Thanks a lot!

Reply:
Sorry for the slow response, does this code build on your machine? I'm trying to build this branch to reproduce the issue.

Reply:
> Sorry for the slow response, does this code build on your machine? I'm trying to build this branch to reproduce the issue.

Sorry was replying late, yes it does.

Reply:
I'm unable to build your branch, I get a bunch of errors from the Concat related tests in ann_layer_test.cpp

Reply:
Oh, well I was just building mlpack. (without tests)
And yes, we'd need to resolve the issue for Concat's Add function, which I had removed.. I'll try to work that out. But, I guess `make mlpack` works for now.

Reply:
So, I tried to create separate sequential layers for the network subparts (in the form of objects), and added them to the FFN(adding their addresses).. and so, while copying, I copy over the Sequential layers(clone the objects), and make a FFN, linking the cloned network object's adddress. (I hope I am using the right terms here :sweat_smile: ).. But all that doesn't seem to work.. I guess there is some error while deleting the network, it happens when one uses `std::move` (I confirmed that with the debugger)..Apart from that the network seems to be working as it should.

Reply:
Yes, with `std::move ` will transfer the ownership, so I guess I can just build mlpack and link against your example?

Reply:
Pretty sure we can figure it out.

Reply:
The code looks good to me, will have to debug the code.

Reply:
> so I guess I can just build mlpack and link against your example?

Yeah, that should work.

Reply:
Also to make sure I get this right the current code works, but at the end it fails since there is some problem with the deconstruction?